File: iOS/PricetarPartner/PricetarPartner/Http/DataModel/UserData.swift
//
//  UserData.swift
//  PricetarPartner
//
//  Created by Leapfrog-Inc. on 2021/02/21.
//

import Foundation

struct UserData {
    
    enum ProfileType: String {
        case none = "0"
        case client = "1"
        case partner = "2"
        
        func toText() -> String {
            switch self {
            case .none:
                return ""
            case .client:
                return "クライアント"
            case .partner:
                return "パートナー"
            }
        }
    }
        
    let id: String
    let nickname: String
    let area: String
    let profileType: ProfileType
    let clientUseFrequency: String
    let clientNewCondition: String
    let clientOldCondition: String
    let clientGenres: [String]
    let clientOptions: [String]
    let clientMessage: String
    let partnerCareer: String
    let partnerStatus: String
    let partnerNewPrice: String
    let partnerOldPrice: String
    let partnerDangerousPrice: String
    let partnerDangerousMessage: String
    let partnerBigPrice: String
    let partnerBigMessage: String
    let partnerInspectionPrice: String
    let partnerInspectionMessage: String
    let partnerMessage: String
    let loginDatetime: Date
    
    init?(data: Dictionary<String, Any>) {
     
        guard let id = data["id"] as? String else {
            return nil
        }
        self.id = id
        
        self.nickname = (data["nickname"] as? String)?.base64Decode() ?? ""
        self.area = (data["area"] as? String)?.base64Decode() ?? ""
        
        guard let profileType = ProfileType(rawValue: data["profileType"] as? String ?? "") else {
            return nil
        }
        self.profileType = profileType
        
        self.clientUseFrequency = (data["clientUseFrequency"] as? String)?.base64Decode() ?? ""
        self.clientNewCondition = (data["clientNewCondition"] as? String)?.base64Decode() ?? ""
        self.clientOldCondition = (data["clientOldCondition"] as? String)?.base64Decode() ?? ""
        self.clientGenres = (data["clientGenres"] as? String)?.components(separatedBy: ",").compactMap { $0.base64Decode() } ?? []
        self.clientOptions = (data["clientOptions"] as? String)?.components(separatedBy: ",").compactMap { $0.base64Decode() } ?? []
        self.clientMessage = (data["clientMessage"] as? String)?.base64Decode() ?? ""
        self.partnerCareer = (data["partnerCareer"] as? String)?.base64Decode() ?? ""
        self.partnerStatus = (data["partnerStatus"] as? String)?.base64Decode() ?? ""
        self.partnerNewPrice = (data["partnerNewPrice"] as? String)?.base64Decode() ?? ""
        self.partnerOldPrice = (data["partnerOldPrice"] as? String)?.base64Decode() ?? ""
        self.partnerDangerousPrice = (data["partnerDangerousPrice"] as? String)?.base64Decode() ?? ""
        self.partnerDangerousMessage = (data["partnerDangerousMessage"] as? String)?.base64Decode() ?? ""
        self.partnerBigPrice = (data["partnerBigPrice"] as? String)?.base64Decode() ?? ""
        self.partnerBigMessage = (data["partnerBigMessage"] as? String)?.base64Decode() ?? ""
        self.partnerInspectionPrice = (data["partnerInspectionPrice"] as? String)?.base64Decode() ?? ""
        self.partnerInspectionMessage = (data["partnerInspectionMessage"] as? String)?.base64Decode() ?? ""
        self.partnerMessage = (data["partnerMessage"] as? String)?.base64Decode() ?? ""
        
        guard let loginDatetimeStr = data["loginDatetime"] as? String, loginDatetimeStr.count == 14, let loginDatetime = DateFormatter(dateFormat: "yyyyMMddHHmmss").date(from: loginDatetimeStr) else {
            return nil
        }
        self.loginDatetime = loginDatetime
    }
    
    static func imageUrl(userId: String) -> String {
        return Constants.ServerRootUrl + "data/user/" + userId + "/image"
    }
    
    func lastLoginString() -> String {
        
        let diff = Date().timeIntervalSince(self.loginDatetime)
        if diff < 60 {
            return "たった今"
        } else if diff < 60 * 60 {
            return "\(Int(floor(diff / 60)))分前"
        } else if diff < 24 * 60 * 60 {
            return "\(Int(floor(diff / 60 / 60)))時間前"
        } else if diff < 30 * 24 * 60 * 60 {
            return "\(Int(floor(diff / 60 / 60 / 24)))日前"
        } else {
            return "1ヶ月以上前"
        }
    }
}

File: Android/app/src/main/java/jp/co/lfg/pricetarpartner/Http/DataModel/UserData.java
package jp.co.lfg.pricetarpartner.Http.DataModel;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Date;

import jp.co.lfg.pricetarpartner.System.Base64Utility;
import jp.co.lfg.pricetarpartner.System.Constants;
import jp.co.lfg.pricetarpartner.System.DateUtility;

public class UserData {

    public enum ProfileType {
        none,
        client,
        partner;

        public static ProfileType create(String value) {

            if (value.equals("0")) {
                return none;
            } else if (value.equals("1")) {
                return client;
            } else if (value.equals("2")) {
                return partner;
            } else {
                return null;
            }
        }

        public String toValue() {

            if (this == none) {
                return "0";
            } else if (this == client) {
                return "1";
            } else {
                return "2";
            }
        }
    }

    public String id;
    public String nickname;
    public String area;
    public ProfileType profileType;
    public String clientUseFrequency;
    public String clientNewCondition;
    public String clientOldCondition;
    public ArrayList<String> clientGenres;
    public String clientMessage;
    public String partnerCareer;
    public String partnerStatus;
    public String partnerNewPrice;
    public String partnerOldPrice;
    public String partnerDangerousPrice;
    public String partnerDangerousMessage;
    public String partnerBigPrice;
    public String partnerBigMessage;
    public String partnerInspectionPrice;
    public String partnerInspectionMessage;
    public String partnerMessage;
    public Date loginDatetime;

    public static UserData create(JSONObject json) {

        try {
            UserData userData = new UserData();

            userData.id = json.getString("id");
            userData.nickname = Base64Utility.decode(json.getString("nickname"));
            userData.area = Base64Utility.decode(json.getString("area"));

            userData.profileType = ProfileType.create(json.getString("profileType"));
            if (userData.profileType == null) {
                return null;
            }

            userData.clientUseFrequency = Base64Utility.decode(json.getString("clientUseFrequency"));
            userData.clientNewCondition = Base64Utility.decode(json.getString("clientNewCondition"));
            userData.clientOldCondition = Base64Utility.decode(json.getString("clientOldCondition"));

            userData.clientGenres = new ArrayList<>();
            String[] clientGenres = json.getString("clientGenres").split(",");
            for (int i = 0; i < clientGenres.length; i++) {
                userData.clientGenres.add(Base64Utility.decode(clientGenres[i]));
            }

            userData.clientMessage = Base64Utility.decode(json.getString("clientMessage"));
            userData.partnerCareer = Base64Utility.decode(json.getString("partnerCareer"));
            userData.partnerStatus = Base64Utility.decode(json.getString("partnerStatus"));
            userData.partnerNewPrice = Base64Utility.decode(json.getString("partnerNewPrice"));
            userData.partnerOldPrice = Base64Utility.decode(json.getString("partnerOldPrice"));
            userData.partnerDangerousPrice = Base64Utility.decode(json.getString("partnerDangerousPrice"));
            userData.partnerDangerousMessage = Base64Utility.decode(json.getString("partnerDangerousMessage"));
            userData.partnerBigPrice = Base64Utility.decode(json.getString("partnerBigPrice"));
            userData.partnerBigMessage = Base64Utility.decode(json.getString("partnerBigMessage"));
            userData.partnerInspectionPrice = Base64Utility.decode(json.getString("partnerInspectionPrice"));
            userData.partnerInspectionMessage = Base64Utility.decode(json.getString("partnerInspectionMessage"));
            userData.partnerMessage = Base64Utility.decode(json.getString("partnerMessage"));

            userData.loginDatetime = DateUtility.stringToDate(json.getString("loginDatetime"), "yyyyMMddkkmmss");
            if (userData.loginDatetime == null) {
                return null;
            }

            return userData;

        } catch (Exception e) {}

        return null;
    }

    public static String getImageUrl(String userId) {
        return Constants.ServerRootUrl + "data/user/" + userId + "/image";
    }

    public String lastLoginString() {

        long diff = ((new Date()).getTime() - this.loginDatetime.getTime()) / 1000;

        if (diff < 60) {
            return "たった今";
        } else if (diff < 60 * 60) {
            return String.valueOf((int)Math.floor(diff / 60)) + "分前";
        } else if (diff < 24 * 60 * 60) {
            return String.valueOf((int)Math.floor(diff / 60 / 60)) + "時間前";
        } else if (diff < 30 * 24 * 60 * 60) {
            return String.valueOf((int)Math.floor(diff / 60 / 60 / 24)) + "日前";
        } else {
            return "1ヶ月以上前";
        }
    }
}
